cmd/cronosd/cmd/fix-unlucky-tx.go

```diff
@@ -0,0 +1,297 @@
+				}
+			}()
+
+			encCfg := app.MakeEncodingConfig()
+
+			appCreator := func() *app.App {
+				return app.New(
+					log.NewNopLogger(), db, nil, false, nil,
+					ctx.Config.RootDir, 0, encCfg, ctx.Viper,
+				)
+			}
+
+			for height := startHeight; height <= endHeight; height++ {
+				blockResult, err := tmDB.stateStore.LoadABCIResponses(height)
+				if err != nil {
+					return err
+				}
+
+				txIndex := findUnluckyTx(blockResult)
+				if txIndex < 0 {
+					// no unlucky tx in the block
+					continue
+				}
+
+				result, err := tmDB.replayTx(appCreator, height, txIndex, state.InitialHeight)
+				if err != nil {
+					return err
+				}
+
+				if err := tmDB.patchDB(blockResult, result); err != nil {
+					return err
+				}
+
+				// decode the tx to get eth tx hashes to log
+				tx, err := clientCtx.TxConfig.TxDecoder()(result.Tx)
+				if err != nil {
+					fmt.Println("can't parse the patched tx", result.Height, result.Index)
+					continue
+				}
+				for _, msg := range tx.GetMsgs() {
+					ethMsg, ok := msg.(*evmtypes.MsgEthereumTx)
+					if ok {
+						fmt.Println("patched", ethMsg.Hash, result.Height, result.Index)
+					}
+				}
+			}
+			return nil
+		},
+	}
+	cmd.Flags().String(flags.FlagChainID, "cronosmainnet_25-1", "network chain ID")
+
+	return cmd
+}
+
+func openAppDB(rootDir string) (dbm.DB, error) {
+	dataDir := filepath.Join(rootDir, "data")
+	return sdk.NewLevelDB("application", dataDir)
+}
+
+type tmDB struct {
+	blockStore *tmstore.BlockStore
+	stateStore sm.Store
+	txIndexer  txindex.TxIndexer
+}
+
+func openTMDB(cfg *tmcfg.Config, chainID string) (*tmDB, error) {
+	// open tendermint db
+	tmdb, err := tmnode.DefaultDBProvider(&tmnode.DBContext{ID: "blockstore", Config: cfg})
+	if err != nil {
+		return nil, err
+	}
+	blockStore := tmstore.NewBlockStore(tmdb)
+
+	stateDB, err := tmnode.DefaultDBProvider(&tmnode.DBContext{ID: "state", Config: cfg})
+	if err != nil {
+		return nil, err
+	}
+	stateStore := sm.NewStore(stateDB)
+
+	txIndexer, err := newTxIndexer(cfg, chainID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &tmDB{
+		blockStore, stateStore, txIndexer,
+	}, nil
+}
+
+func newTxIndexer(config *tmcfg.Config, chainID string) (txindex.TxIndexer, error) {
+	switch config.TxIndex.Indexer {
+	case "kv":
+		store, err := tmnode.DefaultDBProvider(&tmnode.DBContext{ID: "tx_index", Config: config})
+		if err != nil {
+			return nil, err
+		}
+
+		return kv.NewTxIndex(store), nil
+	case "psql":
+		if config.TxIndex.PsqlConn == "" {
+			return nil, errors.New(`no psql-conn is set for the "psql" indexer`)
+		}
+		es, err := psql.NewEventSink(config.TxIndex.PsqlConn, chainID)
+		if err != nil {
+			return nil, fmt.Errorf("creating psql indexer: %w", err)
+		}
+		return es.TxIndexer(), nil
+	default:
+		return nil, fmt.Errorf("unsupported tx indexer backend %s", config.TxIndex.Indexer)
+	}
+}
+
+func findUnluckyTx(blockResult *tmstate.ABCIResponses) int {
+	for txIndex, txResult := range blockResult.DeliverTxs {
```

Comment:
do we need to go through all the deliver txs or can we assume it is always the last that can be false failed?

Reply:
there could be a bunch of out of block gas txs following:
```
"11 out of gas in location: block gas meter; gasWanted: 0, gasUsed: 578156: out of gas"
"11 no block gas left to run tx: out of gas"
"11 no block gas left to run tx: out of gas"
"11 no block gas left to run tx: out of gas"
```